art/attacks/evasion/laser_attack/laser_attack.py

```diff
@@ -0,0 +1,393 @@
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import logging
+from typing import List, Optional, Tuple, Union, Callable
+
+import numpy as np
+from art.attacks.attack import EvasionAttack
+from art.attacks.evasion.laser_attack.algorithms import greedy_search
+from art.attacks.evasion.laser_attack.utils import (AdversarialObject,
+                                                    AdvObjectGenerator,
+                                                    DebugInfo, ImageGenerator,
+                                                    Line, wavelength_to_RGB)
+
+logger = logging.getLogger(__name__)
+
+class LaserAttack(EvasionAttack):
+    attack_params = EvasionAttack.attack_params + [
+        "iterations",
+        "laser_generator",
+        "image_generator",
+        "random_initializations",
+        "optimisation_algorithm",
+        "debug",
+    ]
+    _estimator_requirements = ()
+
+    def __init__(
+        self,
+        estimator,
+        iterations: int,
+        laser_generator: AdvObjectGenerator,
+        image_generator: ImageGenerator = ImageGenerator(),
+        random_initializations: int = 1,
+        optimisation_algorithm: Callable = greedy_search,
+        tensor_board: Union[str, bool] = False,
+        debug: Optional[DebugInfo] = None
+    ) -> None:
+        """
+        :param estimator: Predictor of the image class.
+        :param iterations: Maximum number of iterations of the algorithm.
+        :param laser_generator: Object responsible for
+            generation laser beams images and their updation.
+        :param image_generator: Object responsible for image generation.
+        :param random_initializations: How many times repeat the attack.
+        :param tensor_board:
+        :param debug: Optional debug handler.
+        """
+
+        super().__init__(estimator=estimator, tensor_board=tensor_board)
+        self.iterations = iterations
+        self.random_initializations = random_initializations
+        self.optimisation_algorithm = optimisation_algorithm
+        self._laser_generator = laser_generator
+        self._image_generator = image_generator
+        self._debug = debug
+
+        self._check_params()
+
+    def generate(
+        self,
+        x: np.ndarray,
+        *args,
+        **kwargs
+    ) -> Optional[List]:
+        """
+
+        Generate adversarial example for a single image.
+
+        :param x: Image to attack as a tensor (NRGB = (1, ...))
+        :return: List of paris of adversarial objects and predicted class.
+        """
+
+        adversarial_params = []
+        for image_index in range(x.shape[0]):
+            params, adv_class = self._generate_for_single_input(x[image_index])
+            adversarial_params.append((params, adv_class))
+
+        return adversarial_params
```

Comment:
Method `generate` should return adversarial examples `x_adv` of the same shape as `x` and of type `np.ndarray`.